attest/verifier/src/ias.rs

```diff
@@ -0,0 +1,244 @@
+                            && Certificate::verify_with_profile(
+                                &existing_cert_single,
+                                &chain_cert_single,
+                                None,
+                                Some(&profile),
+                                None,
+                            )
+                            .is_ok()
+                        {
+                            signer_chain.push(chain_cert);
+                            continue 'outer;
+                        }
+                    }
+
+                    break;
+                }
+                Some(signer_chain)
+            })
+            // Then see if any of those chains are connected to a trust anchor
+            .find_map(|mut signer_chain| {
+                let signer_toplevel = signer_chain
+                    .last()
+                    .expect("Signer chain was somehow emptied before use.");
+                let mut signer_toplevel_single = MbedtlsList::new();
+                signer_toplevel_single.push(signer_toplevel.clone());
+
+                // First, check if the last element in the chain is signed by a trust anchor
+                for cacert in &trust_anchors {
+                    let mut cacert_single = MbedtlsList::new();
+                    cacert_single.push(cacert.clone());
+                    if Certificate::verify_with_profile(
+                        &signer_toplevel_single,
+                        &cacert_single,
+                        None,
+                        Some(&profile),
+                        None,
+                    )
+                    .is_ok()
+                    {
+                        return Some(());
+                    }
+                }
+
+                // Otherwise, check if any of the pubkeys in the chain are a trust anchor
+                for cert in &mut signer_chain {
+                    for cacert in &mut trust_anchors {
+                        if cert.public_key_mut().write_public_der_vec()
+                            == cacert.public_key_mut().write_public_der_vec()
+                        {
+                            return Some(());
+                        }
+                    }
+                }
+                None
+            })
+            .ok_or(Error::BadSignature)?;
+
+        let report_data = VerificationReportData::try_from(report)?;
+
+        if (self.and_verifiers.is_empty()
+            || self
+                .and_verifiers
+                .iter()
+                .all(|verifier| verifier.verify(&report_data)))
+            && (self.or_verifiers.is_empty()
+                || self
+                    .or_verifiers
+                    .iter()
+                    .any(|verifier| verifier.verify(&report_data)))
+        {
+            Ok(report_data)
+        } else {
+            Err(Error::Verification(report_data))
+        }
+    }
```

Comment:
This function does seem a bit too long. Also, the nested loops are too complex. Maybe break up the code a bit?

Reply:
The signature chain validation will be moved to it's own "signature verifier" type in phase 4 of this, but the signature verification is complex because of the reasons discussed in the comment.

The chained iterator scheme is also meant to reduce the number of times we're doing a heap allocation to the bare minimum, and `clippy` (which is a hard CI requirement) complains if you're creating intermediate collections just to iterate over them again, which would be the obvious way to make this more readable.